src/main/java/com/jcabi/s3/AwsListIterator.java

```diff
@@ -0,0 +1,154 @@
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.jcabi.s3;
+
+import com.amazonaws.AmazonServiceException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.jcabi.log.Logger;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+
+/**
+ * Iterator for large lists returned by S3.
+ * @author Roman Kisilenko (roman.kisilenko@gmail.com)
+ * @version $Id$
+ * @since 0.10
+ */
+class AwsListIterator implements Iterator<String> {
+
+    /**
+     * Region we're in.
+     */
+    private final transient Region region;
+
+    /**
+     * Bucket name.
+     */
+    private final transient String bucket;
+
+    /**
+     * Prefix to use.
+     */
+    private final transient String prefix;
+
+    /**
+     * Partial S3 response iterator.
+     */
+    private transient Iterator<String> partial;
+
+    /**
+     * Next marker for partial response or null if response contains no more
+     * data.
+     */
+    private transient String marker;
+
+    /**
+     * Constructs AwsListIterator.
+     * @param pfx Key prefix
+     * @param rgn Region we're in
+     * @param bkt Bucket name
+     */
+    public AwsListIterator(final Region rgn, final String bkt,
+        final String pfx) {
+        this.prefix = pfx;
+        this.region = rgn;
+        this.bucket = bkt;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final boolean hasNext() {
+        if (this.partial == null || (!this.partial.hasNext()
+            && this.marker != null)) {
+            this.load();
+        }
+        return this.partial.hasNext();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final String next() {
+        if (!this.hasNext()) {
+            throw new NoSuchElementException(
+                "There are no more elements in this iterator"
+            );
+        }
+        return this.partial.next();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void remove() {
+        throw new UnsupportedOperationException("Remove is not supported");
+    }
+
+    /**
+     * Loads next portion of data from S3.
+     */
+    private void load() {
```

Comment:
Would be much more elegant to **return** a list from this method and save it into `this.partial` in the method above. Now this method looks like a self-modifying procedure, which is rather error-prone, since it's not clear what other side-effects if may have.